### DISCUSSION TITLE
Welcome to JAX-in-Cell Discussions!

### DISCUSSION THREAD

**Discussion by rogeriojorge** [uwplasma/JAX-in-Cell#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by LPI-MF** [comment]
Hi, this is Mufei, a postdoc from Oxford. Thank you for sharing this interesting PIC code. I am pretty curious about how this Python-based code can be combined with ML.

Best wishes
Mufei

**Comment by rogeriojorge** [comment]
Hi @LPI-MF, thank you for reaching out! The connection with ML is its differentiability. 

A differentiable forward model such as JAX-in-Cell can be coupled to machine learning by placing the simulator directly in the training loop. This is only possible due to its automatic differentiation capabilities, where we can obtain gradients of task-relevant objectives (diagnostics and losses) with respect to inputs, parameters, and embedded model components. We can then do gradient-based optimization and learn over high-dimensional parameter spaces without the instability and cost of finite-difference derivatives—so gradients can propagate end-to-end through any workflow. This allows us to perform tasks such as parameter inference and calibration from experimental diagnostics, data assimilation, optimal design of actuators/geometry/profiles, model-based control, and end-to-end training of hybrid physics–ML components such as closures or subgrid models, all while optimizing the actual objective rather than proxy quantities.

**Reply by LPI-MF** [reply]
Thank you for the reply. I see that so in this differentiable PIC framework, simulation inputs and model parameters can be treated as trainable variables analogous to ML weights. Automatic differentiation enables stable, end-to-end gradient propagation through the simulator, allowing these parameters to be optimized directly against task-relevant objectives within a training loop. I am pretty interested in using this code to learn about the collision operator. I imagine another ML model can be coupled to the PIC as well.

**Comment by liyongxin0000** [comment]
Hi,

I want to add a laser field to my simulation. I tried setting external_electric_field_amplitude in input.toml, but it doesn't seem to work (the code seems commented out in _simulation.py).

I also tried passing an external_electric_field array manually, but it behaves like a static field, not a propagating wave.

How can I correctly inject a propagating electromagnetic wave (laser) in JAX-in-Cell?

Thanks!

**Reply by Longyu975** [reply]
Hi @liyongxin0000. Great to hear that you’re getting use out of our code. We haven’t benchmarked the laser implementation yet, and there is still some work to be done. Here are two potential approaches you could consider for your use case.

At the moment, the external field is static and applied only in the x direction. If you’d like a time-dependent field, one simple approach is to add a modulation directly in **algorithm.py**, for example, multiplying by sin(ω * step_index * dt) to make the field oscillatory. For more complex laser profiles, you could define your own function, such as a propagating wave sin(kx + ωt) in the y–z directions.

An alternative to manually adding fields is to modify **boundary_conditions.py**, similar to how EPOCH handles lasers (see: https://github.com/epochpic/epoch/blob/main/epoch1d/src/laser.f90). In this approach, you use absorbing (i.e. no-reflection) boundary conditions and add a laser source at the boundary, which then evolves freely under Maxwell’s equations at each simulation step.

**Reply by liyongxin0000** [reply]
Thank you for the detailed explanation! I appreciate your help and suggestions.

**Comment by thenitinshukla** [comment]
Hi @rogeriojorge, When you mention CPU and GPU (according to doc), are you referring to a single CPU and a single GPU? Just to confirm, the code does not use multiple nodes or multiple GPUs, correct? Thanks!

**Reply by rogeriojorge** [reply]
Hi @thenitinshukla, as JAX-in-Cell is only 1D3D, and parallelization in PIC codes is usually done in the additional space dimensions, it now runs on single CPUs and GPUs. In fact, it is extremely optimized to run on those, making use of the XLA compiler optimizations for single CPU/GPU.
